sim/src/edge.rs

```diff
@@ -63,8 +63,8 @@ impl SimElement for Edge {
         assert!(self.neighbors.len() < 2);
         self.neighbors.push(neighbor);
     }
-    fn whoami(&self) -> (&str, &Vec<String>) {
-        return (&self.id, &self.neighbors);
+    fn whoami(&self) -> (&str, &Vec<String>, Option<&str>) {
```

Comment:
This function is a little overloaded. There should actually be three functions.
`whoami `to return the name/id of the node, `get_neighbors `to retrieve the neighbors. The last thing should be an operation specific to storage, if possible.

Reply:
I agree it would make sense to be specific to storage, but I'm not sure how that would work.  Because all the elements are in Box<dyn PrintableElement>, there is no way to tell what the types of your elements actually are.  And the sort of type-checking you can do in other languages doesn't seem to be available in Rust (see: https://users.rust-lang.org/t/how-can-i-print-the-type-of-a-variable/4183/6 and the description here: https://doc.rust-lang.org/core/any/fn.type_name.html

So I can either make storage some sort of special vector of storages that the simulator treats differently from all other elements.  This would muddy up how we send RPCs on, and muddy up the overall structure of the simulator quite a bit.  I could also just have an extra function that only storage uses in sim_element.  I can't think of another solution, so I went with the latter.

Reply:
Let me think about this for a bit. If we have to add a new function to the sim interface every time we add some new subnode with different functionality, we will get into a big mess pretty soon. So we must be doing something wrong here. There must be a cleaner way to do this in Rust.

Reply:
One thing that already bothers me is that they are all Boxed.

Reply:
Why are the boxes undesirable?

Reply:
Well for one the reason you gave and the fact that they are heap allocated. Just need to understand why this is necessary.

Reply:
My impression is that it's heap allocated so the compiler doesn't need to worry about the size of its HashMap of PrintableElements at compile time.